### ISSUE TITLE
Problems using a Python layer: error code -6

### ISSUE THREAD

**Issue by MRLoghmani** [NVIDIA/DIGITS#1702]
I was interested in creating my own layer for data augmentation, so I started following the tutorial about training LeNet-5 on MNIST with square occlusion patches (https://github.com/NVIDIA/DIGITS/tree/master/examples/python-layer).
I followed the instructions step by step, but when I create the model (/start the training) it stops immediately with error code -6. I didn't manage to find anything useful about this problem. Any suggestions?

This is the screenshot from DIGITS:
![mnist_log](https://user-images.githubusercontent.com/6833023/27679356-c45cd89e-5cb8-11e7-86e4-3ecd0bebe4d5.png)

**Comment by lukeyeager** [comment]
Click on the "Raw caffe output" file and see if there's anything interesting at the end of it.

**Comment by bdmccord** [comment]
I would also add that, in my experience, python layer errors tend to get swallowed somewhere, so you might want to add a try/except block to your layer and put this in your except block:

```		except Exception as e:
			traceback.print_exc()
			raise e
```

**Comment by MRLoghmani** [comment]
I checked the "caffe_output.log" file and it seems the problem is the version ob libprotoc.

""""
[libprotobuf FATAL google/protobuf/stubs/common.cc:61] This program requires version 3.2.0 of the Protocol Buffer runtime library, but the installed version is 2.6.1.  Please update your library.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "google/protobuf/descriptor.pb.cc".)
terminate called after throwing an instance of 'google::protobuf::FatalException'
what():  This program requires version 3.2.0 of the Protocol Buffer runtime library, but the installed version is 2.6.1.  Please update your library.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "google/protobuf/descriptor.pb.cc".)
""""

I updated it to version 3.2.0, but it seems that caffe is not properly linked to it because the problem persists. Re-compiling caffe didn't help.

**Comment by MRLoghmani** [comment]
I solved this problem in 4 steps:

1) remove old libprotobuf:
$ sudo apt-get --purge remove libprotobuf*

2) build libprotobuf 3.2.0 from source: https://github.com/google/protobuf/tree/master/src

3) insert /usr/local/lib on top of my LD_LIBRARY_PATH variable:
$ export LD_LIBRARY_PATH=/usr/local/lib:$LD_LIBRARY_PATH

4) re-compile Caffe


I then encountered a different problem with error code -11:

""""
I0703 13:42:24.320775 16248 upgrade_proto.cpp:1082] Attempting to upgrade input file specified using deprecated 'solver_type' field (enum)': /mnt/4TB_HDD/digits-jobs/20170703-134222-5c27/solver.prototxt
I0703 13:42:24.320991 16248 upgrade_proto.cpp:1089] Successfully upgraded file specified using deprecated 'solver_type' field (enum) to 'type' field (string).
W0703 13:42:24.320999 16248 upgrade_proto.cpp:1091] Note that future Caffe releases will only support 'type' field (string) for a solver's type.
I0703 13:42:24.321089 16248 caffe.cpp:218] Using GPUs 0
I0703 13:42:24.329012 16248 caffe.cpp:223] GPU 0: TITAN X (Pascal)
I0703 13:42:24.706733 16248 solver.cpp:44] Initializing solver from parameters:
test_iter: 469
test_interval: 704
base_lr: 0.01
display: 79
max_iter: 21120
lr_policy: "step"
gamma: 0.1
momentum: 0.9
weight_decay: 0.0001
stepsize: 6970
snapshot: 704
snapshot_prefix: "snapshot"
solver_mode: GPU
device_id: 0
net: "train_val.prototxt"
train_state {
level: 0
stage: ""
}
type: "SGD"
I0703 13:42:24.706851 16248 solver.cpp:87] Creating training net from net file: train_val.prototxt
I0703 13:42:24.707056 16248 net.cpp:294] The NetState phase (0) differed from the phase (1) specified by a rule in layer val-data
I0703 13:42:24.707065 16248 net.cpp:294] The NetState phase (0) differed from the phase (1) specified by a rule in layer accuracy
I0703 13:42:24.707120 16248 net.cpp:51] Initializing net from parameters:
state {
phase: TRAIN
level: 0
stage: ""
}
layer {
name: "train-data"
type: "Data"
top: "data"
top: "label"
include {
phase: TRAIN
}
transform_param {
mean_file: "/mnt/4TB_HDD/digits-jobs/20170629-103511-e443/mean.binaryproto"
}
data_param {
source: "/mnt/4TB_HDD/digits-jobs/20170629-103511-e443/train_db"
batch_size: 64
backend: LMDB
}
}
layer {
name: "scale"
type: "Power"
bottom: "data"
top: "scaled"
power_param {
scale: 0.0125
}
}
layer {
name: "blank_square"
type: "Python"
bottom: "scaled"
top: "scaled"
python_param {
module: "digits_python_layers"
layer: "BlankSquareLayer"
}
}
layer {
name: "conv1"
type: "Convolution"
bottom: "scaled"
top: "conv1"
param {
lr_mult: 1
}
param {
lr_mult: 2
}
convolution_param {
num_output: 20
kernel_size: 5
stride: 1
weight_filler {
type: "xavier"
}
bias_filler {
type: "constant"
}
}
}
layer {
name: "pool1"
type: "Pooling"
bottom: "conv1"
top: "pool1"
pooling_param {
pool: MAX
kernel_size: 2
stride: 2
}
}
layer {
name: "conv2"
type: "Convolution"
bottom: "pool1"
top: "conv2"
param {
lr_mult: 1
}
param {
lr_mult: 2
}
convolution_param {
num_output: 50
kernel_size: 5
stride: 1
weight_filler {
type: "xavier"
}
bias_filler {
type: "constant"
}
}
}
layer {
name: "pool2"
type: "Pooling"
bottom: "conv2"
top: "pool2"
pooling_param {
pool: MAX
kernel_size: 2
stride: 2
}
}
layer {
name: "ip1"
type: "InnerProduct"
bottom: "pool2"
top: "ip1"
param {
lr_mult: 1
}
param {
lr_mult: 2
}
inner_product_param {
num_output: 500
weight_filler {
type: "xavier"
}
bias_filler {
type: "constant"
}
}
}
layer {
name: "relu1"
type: "ReLU"
bottom: "ip1"
top: "ip1"
}
layer {
name: "ip2"
type: "InnerProduct"
bottom: "ip1"
top: "ip2"
param {
lr_mult: 1
}
param {
lr_mult: 2
}
inner_product_param {
num_output: 10
weight_filler {
type: "xavier"
}
bias_filler {
type: "constant"
}
}
}
layer {
name: "loss"
type: "SoftmaxWithLoss"
bottom: "ip2"
bottom: "label"
top: "loss"
}
I0703 13:42:24.707165 16248 layer_factory.hpp:77] Creating layer train-data
I0703 13:42:24.707263 16248 db_lmdb.cpp:35] Opened lmdb /mnt/4TB_HDD/digits-jobs/20170629-103511-e443/train_db
I0703 13:42:24.707283 16248 net.cpp:84] Creating Layer train-data
I0703 13:42:24.707291 16248 net.cpp:380] train-data -> data
I0703 13:42:24.707307 16248 net.cpp:380] train-data -> label
I0703 13:42:24.707316 16248 data_transformer.cpp:25] Loading mean file from: /mnt/4TB_HDD/digits-jobs/20170629-103511-e443/mean.binaryproto
I0703 13:42:24.708803 16248 data_layer.cpp:45] output data size: 64,1,28,28
I0703 13:42:24.710686 16248 net.cpp:122] Setting up train-data
I0703 13:42:24.710697 16248 net.cpp:129] Top shape: 64 1 28 28 (50176)
I0703 13:42:24.710700 16248 net.cpp:129] Top shape: 64 (64)
I0703 13:42:24.710701 16248 net.cpp:137] Memory required for data: 200960
I0703 13:42:24.710707 16248 layer_factory.hpp:77] Creating layer scale
I0703 13:42:24.710717 16248 net.cpp:84] Creating Layer scale
I0703 13:42:24.710721 16248 net.cpp:406] scale <- data
I0703 13:42:24.710729 16248 net.cpp:380] scale -> scaled
I0703 13:42:24.710752 16248 net.cpp:122] Setting up scale
I0703 13:42:24.710757 16248 net.cpp:129] Top shape: 64 1 28 28 (50176)
I0703 13:42:24.710757 16248 net.cpp:137] Memory required for data: 401664
I0703 13:42:24.710759 16248 layer_factory.hpp:77] Creating layer blank_square
*** Aborted at 1499082145 (unix time) try "date -d @1499082145" if you are using GNU date ***
PC: @     0x7f4639a0d873 std::_Hashtable<>::clear()
*** SIGSEGV (@0x9) received by PID 16248 (TID 0x7f46717e1ac0) from PID 9; stack trace: ***
@     0x7f466f64d4b0 (unknown)
@     0x7f4639a0d873 std::_Hashtable<>::clear()
@     0x7f46399ff346 google::protobuf::DescriptorPool::FindFileByName()
@     0x7f46399ddac8 google::protobuf::python::cdescriptor_pool::AddSerializedFile()
@     0x7f466139c7d0 PyEval_EvalFrameEx
@     0x7f46614c501c PyEval_EvalCodeEx
@     0x7f466141b3dd (unknown)
@     0x7f46613ee1e3 PyObject_Call
@     0x7f466140eae5 (unknown)
@     0x7f46613a5123 (unknown)
@     0x7f46613ee1e3 PyObject_Call
@     0x7f466139913c PyEval_EvalFrameEx
@     0x7f46614c501c PyEval_EvalCodeEx
@     0x7f4661393b89 PyEval_EvalCode
@     0x7f46614281b4 PyImport_ExecCodeModuleEx
@     0x7f4661428b8f (unknown)
@     0x7f466142a300 (unknown)
@     0x7f466142a5c8 (unknown)
@     0x7f466142b6db PyImport_ImportModuleLevel
@     0x7f46613a2698 (unknown)
@     0x7f46613ee1e3 PyObject_Call
@     0x7f46614c4447 PyEval_CallObjectWithKeywords
@     0x7f46613975c6 PyEval_EvalFrameEx
@     0x7f46614c501c PyEval_EvalCodeEx
@     0x7f4661393b89 PyEval_EvalCode
@     0x7f46614281b4 PyImport_ExecCodeModuleEx
@     0x7f4661428b8f (unknown)
@     0x7f466142a300 (unknown)
@     0x7f466142a5c8 (unknown)
@     0x7f466142b6db PyImport_ImportModuleLevel
@     0x7f46613a2698 (unknown)
@     0x7f46613ee1e3 PyObject_Call
""""

Any idea?

**Comment by MRLoghmani** [comment]
FYI: I solved the problem thanks to this "issue": https://github.com/BVLC/caffe/issues/5357

In particular, after following the steps described above and re-building caffe with the option WITH_PYTHON_LAYER=1, I run the command:
$ pip install --user --upgrade protobuf==3.1.0.post1
